source/common/network/happy_eyeballs_connection_impl.cc

```diff
@@ -0,0 +1,528 @@
+      []() -> void { ASSERT(false); });
+  if (per_connection_state_.buffer_limits_.has_value()) {
+    post_connect_state_.write_buffer_.value()->setWatermarks(
+        per_connection_state_.buffer_limits_.value());
+  }
+  post_connect_state_.write_buffer_.value()->move(data);
+  post_connect_state_.end_stream_ = end_stream;
+}
+
+void HappyEyeballsConnectionImpl::setBufferLimits(uint32_t limit) {
+  if (!connect_finished_) {
+    ASSERT(!per_connection_state_.buffer_limits_.has_value());
+    per_connection_state_.buffer_limits_ = limit;
+    if (post_connect_state_.write_buffer_.has_value()) {
+      post_connect_state_.write_buffer_.value()->setWatermarks(
+          per_connection_state_.buffer_limits_.value());
+    }
+  }
+  for (auto& connection : connections_) {
+    connection->setBufferLimits(limit);
+  }
+}
+
+uint32_t HappyEyeballsConnectionImpl::bufferLimit() const {
+  if (!connect_finished_) {
+    if (per_connection_state_.buffer_limits_.has_value()) {
+      return per_connection_state_.buffer_limits_.value();
+    }
+    return 0;
+  }
+  return connections_[0]->bufferLimit();
+}
+
+bool HappyEyeballsConnectionImpl::aboveHighWatermark() const {
+  if (!connect_finished_) {
+    return above_write_high_water_mark_;
+  }
+
+  return connections_[0]->aboveHighWatermark();
+}
+
+const ConnectionSocket::OptionsSharedPtr& HappyEyeballsConnectionImpl::socketOptions() const {
+  // Note, this might change before connect finishes.
+  return connections_[0]->socketOptions();
+}
+
+absl::string_view HappyEyeballsConnectionImpl::requestedServerName() const {
+  // Note, this might change before connect finishes.
+  return connections_[0]->requestedServerName();
+}
+
+StreamInfo::StreamInfo& HappyEyeballsConnectionImpl::streamInfo() {
+  // Note, this might change before connect finishes.
+  return connections_[0]->streamInfo();
+}
+
+const StreamInfo::StreamInfo& HappyEyeballsConnectionImpl::streamInfo() const {
+  // Note, this might change before connect finishes.
+  return connections_[0]->streamInfo();
+}
+
+absl::string_view HappyEyeballsConnectionImpl::transportFailureReason() const {
+  // Note, this might change before connect finishes.
+  return connections_[0]->transportFailureReason();
+}
+
+bool HappyEyeballsConnectionImpl::startSecureTransport() {
+  if (!connect_finished_) {
+    per_connection_state_.start_secure_transport_ = true;
+  }
+  bool ret = false;
+  for (auto& connection : connections_) {
+    ret = connection->startSecureTransport();
+  }
+  return ret;
+}
+
+absl::optional<std::chrono::milliseconds> HappyEyeballsConnectionImpl::lastRoundTripTime() const {
+  // Note, this might change before connect finishes.
+  return connections_[0]->lastRoundTripTime();
+}
+
+void HappyEyeballsConnectionImpl::addConnectionCallbacks(ConnectionCallbacks& cb) {
+  if (connect_finished_) {
+    connections_[0]->addConnectionCallbacks(cb);
```

Comment:
Hm, I would have expected that we'd defer adding these until connected, and capture callbacks in this class and propogate up where it makes sense. Mind filling me in on why you went this route?

Reply:
Hm, I'm not sure I understand. The currently implementation *does* defer adding these until connection completes. 
```
  post_connect_state_.connection_callbacks_.push_back(&cb);
```
And then in `onEvent()` they're added to the connection. So I think it does what you're suggesting?

But maybe your thinking that the connection's connectionCallbacks *always* delegate up to HappyEyeballsConnectionImpl and the impl itself passes events from the connection to the caller. Is that what you're thinking? I did consider that but kinda wanted to get the HappyEyeballsConnectionImpl out of the loop once the connection completed. I was thinking about making sure I replicated the right behavior if there we multiple callbacks and if something happened mid iteration. Is there any complexity there? But I'm not remotely passionate about this ... it just kinda happened this way. If this is what you recommend I'd be happy to switch!